pallets/payment-streams/src/utils.rs

```diff
@@ -0,0 +1,490 @@
+                Some(payment_stream) => {
+                    payment_stream.rate = new_rate;
+                    Ok(())
+                }
+                None => {
+                    // This should never happen as we already checked that the payment stream exists
+                    return Err(Error::<T>::PaymentStreamNotFound);
+                }
+            }
+        })?;
+
+        Ok(())
+    }
+
+    /// This function holds the logic that checks if a payment stream can be removed and, if so, removes the payment stream from the PaymentStreams mapping,
+    /// decreases the user's payment streams count and, if the user has no more payment streams, releases the deposit.
+    pub fn do_delete_payment_stream(
+        bsp_account: &T::AccountId,
+        user_account: &T::AccountId,
+    ) -> DispatchResult {
+        // Get the BSP ID of the BSP account
+        let bsp_id = <T::Providers as storage_hub_traits::ProvidersInterface>::get_provider(
+            bsp_account.clone(),
+        )
+        .ok_or(Error::<T>::NotABackupStorageProvider)?;
+
+        // Check that a payment stream between that BSP and user exists
+        ensure!(
+            PaymentStreams::<T>::contains_key(bsp_id, user_account),
+            Error::<T>::PaymentStreamNotFound
+        );
+
+        // Charge the payment stream before deletion to make sure the storage provided by the provider is paid in full for its duration
+        let amount_charged = Self::do_charge_payment_stream(&bsp_account, user_account)?;
+        if amount_charged > Zero::zero() {
+            // We emit a payment charged event only if the user had to pay for its storage before deleting the payment stream
+            Self::deposit_event(Event::<T>::PaymentStreamCharged {
+                user_account: user_account.clone(),
+                backup_storage_provider_id: bsp_id,
+                amount: amount_charged,
+            });
+        }
+
+        // TODO: What do we do when a user is flagged as without funds? Does the provider assume the loss and we remove the payment stream?
+        // Check that the user is not flagged as without funds (after charging, to make sure it makes sense to delete the payment stream)
+        ensure!(
+            !UsersWithoutFunds::<T>::contains_key(user_account),
+            Error::<T>::UserWithoutFunds
+        );
+
+        // Remove the payment stream from the PaymentStreams mapping
+        PaymentStreams::<T>::remove(bsp_id, user_account);
+
+        // Decrease the user's payment streams count
+        let mut user_payment_streams_count = RegisteredUsers::<T>::get(user_account);
+        user_payment_streams_count = user_payment_streams_count
+            .checked_sub(1)
+            .ok_or(ArithmeticError::Underflow)?;
+        RegisteredUsers::<T>::insert(user_account, user_payment_streams_count);
+
+        // If the user has no more payment streams, release the deposit
+        if user_payment_streams_count == 0 {
```

Comment:
Here too, why not using a deposit amount that is proportional to the payment streams a user has?